cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc

```diff
@@ -255,6 +255,17 @@ struct Day {
   Localizer localizer_;
 };
 
+template <typename Duration, typename Localizer>
+struct Day<Duration, Localizer, DurationType> {
```

Comment:
This specialization is necessary because extracting day from duration is different than from timestamp/date: `day(date/timestamp)` returns the "day" component which is in the range of [1, 31], whereas `day(duration)` returns the number of days represented in a specified amount of sec/ms/us/ns whose range is logically [int64_min, int64_max] depending on the unit.